includes/class-wc-stripe-webhook-handler.php

```diff
@@ -238,9 +238,9 @@ public function process_webhook_payment( $notification, $retry = true ) {
 					$order->save();
 				}
 
-				if ( $this->is_no_such_token_error( $response->error ) && $prepared_source->token_id ) {
+				if ( $this->is_no_such_token_error( $response->error ) && $source_object->token_id ) {
```

Comment:
This `if` is always false, since `$source_object->token_id` is only assigned once:

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/a33a56d230e5ab3689a2a8a1926cecbeecb36155/includes/class-wc-stripe-webhook-handler.php#L224-L224

Maybe it should have been removed when the `$prepared_source` variable was?

Reply:
@diegocurbelo I started looking into this further, and there's never a token assigned within the method at all. This means the whole section of code there is likely never going to do anything. Since this is working off of a webhook, the only place a token would be coming from is the order, which is passed to [`generate_payment_request`](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/5.1.0/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L336-L416). That method doesn't look for or use `token_id`, it uses the source that is passed.

On top of that, I found [another section of code](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/5.1.0/includes/class-wc-stripe-order-handler.php#L137-L144) with the same issue with the wrong variable name ( `prepared_source` vs `source_object` ), and it too never has a token assigned.

I am wondering if these should be looking for a [source error](https://stripe.com/docs/error-codes#resource-missing) in the response, then looking for a WC token related to that source and removing it. When looking into [`prepare_source`](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/5.1.0/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L611-L701), it looks like tokens refer to WC tokens only, which are never passed back and forth to Stripe, just sources are.

Reply:
@jessepearson The idea of this `if` block is to remove the payment method from WC if it's not valid anymore in Stripe, we should keep the functionality.

In the WC database we store the `source_id` of the token, so we can get all the order's tokens with `WC_Payment_Tokens::get_order_tokens( $order_id )`... and find the one with the `source_id` from the notification (`$source_id = $notification->data->object->id`)